proxy/src/control/cache.rs

```diff
@@ -71,20 +72,22 @@ impl<T> Cache<T> where T: Clone + Copy + Eq + std::hash::Hash {
             extend_inner(&mut self.values, iter, on_change);
         } else {
             let to_insert = iter.collect::<HashSet<T>>();
-            extend_inner(&mut self.values, to_insert.iter().map(|value| *value), on_change);
+            extend_inner(&mut self.values, to_insert.iter().cloned(), on_change);
             self.retain(&to_insert, on_change);
         }
         self.reset_on_next_modification = false;
     }
 
-    pub fn remove<I, F>(&mut self, iter: I, on_change: &mut F)
-        where I: Iterator<Item = T>,
+    pub fn remove<I, F, Q>(&mut self, iter: I, on_change: &mut F)
+        where I: Iterator<Item = Q>,
+              T: Borrow<Q>,
+              Q: ::std::hash::Hash + Eq,
               F: FnMut(T, CacheChange)
     {
         if !self.reset_on_next_modification {
             for value in iter {
-                if self.values.remove(&value) {
-                    on_change(value, CacheChange::Removal);
+                if let Some(removed) = self.values.take(&value) {
```

Comment:
I don't quite understand how this works...

- We have a `Self`-typed cache holding `T`-typed values.
- We have an `I`-typed iterator of `Q-typed` values.
- `T`-typed values are dereferenced to `Q`-typed values.
- We iterate over `Q`-typed values, and lookup each `Q`-typed value--by reference--into the cache.

How can we do Q-typed lookups for T-typed values? Are we relying on the fact that inserts use take a reference before computing the hash or something? I'm confused.

Reply:
> `T`-typed values are dereferenced to `Q`-typed values.

This is not quite correct; the `T: Borrow<Q>` bound means that `Q`-typed values _are a borrowed form_ of `T`-typed values: https://doc.rust-lang.org/std/borrow/trait.Borrow.HTML

`HashSet` lookups for a `HashSet<T>` don't take arguments of type `T`, they [take arguments of a type `Q` bounded with `T: Borrow<Q>`](https://doc.rust-lang.org/std/collections/struct.HashSet.html#method.take). This is just applying the same bound placed on the `HashSet` lookup functions to the type of the items in the iterator.